File: altgraph_data_app/src/main/java/org/cjoakim/cosmos/altgraph/data/common/graph/v2/JStarNetwork.java
package org.cjoakim.cosmos.altgraph.data.common.graph.v2;

import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.cjoakim.cosmos.altgraph.data.common.ImdbConstants;
import org.cjoakim.cosmos.altgraph.data.common.graph.v2.struct.EdgeStruct;
import org.cjoakim.cosmos.altgraph.data.common.graph.v2.struct.EdgesStruct;
import org.jgrapht.graph.DefaultEdge;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

/**
 * Instances of this class represents a network that starts from one given
 * vertex, and traverses out from that root vertex one "degree of separation"
 * at a time.  The resulting network shape looks like a "star" with the root
 * vertex in the center.
 * <p>
 * Each "degree" (as in the 6-degrees of Kevin Bacon) is represented as
 * instance of JStarDegree.
 * <p>
 * Chris Joakim, Microsoft, November 2022
 */

@Data
@Slf4j
public class JStarNetwork implements ImdbConstants {

    private String rootVertex;
    private int depth;
    private long startMs;
    private long finishMs;

    private ArrayList<JStarDegree> degrees;
    private HashSet<String> visitedSet;
    private HashSet<String> unvisitedSet;

    public JStarNetwork(String root, int depth) {
        super();
        this.startMs = System.currentTimeMillis();
        this.rootVertex = root;
        this.depth = depth;
        this.degrees = new ArrayList<JStarDegree>();
        for (int i = 1; i <= depth; i++) {
            degrees.add(new JStarDegree(i));
        }
        this.visitedSet = new HashSet<String>();
        this.unvisitedSet = new HashSet<String>();
        unvisitedSet.add(rootVertex);
    }

    public void addOutEdgesFor(String vertex, Set<DefaultEdge> outEdges, int degree) {
        visitedSet.add(vertex);
        degrees.get(degree - 1).add(vertex, outEdges);

        Iterator<DefaultEdge> it = outEdges.iterator();
        while (it.hasNext()) {
            DefaultEdge de = it.next();
            String[] values = parseDefaultEdge(de);
            String ev1 = values[0];
            String ev2 = values[1];
            if (!visitedSet.contains(ev1)) {
                unvisitedSet.add(ev1);
            }
            if (!visitedSet.contains(ev2)) {
                unvisitedSet.add(ev2);
            }
        }
    }

    private String[] parseDefaultEdge(DefaultEdge de) {
        String[] result = new String[2];
        result[0] = "?".intern();
        result[1] = "?".intern();

        if (de != null) {
            String[] tokens = de.toString().split(":");
            if (tokens.length == 2) {
                result[0] = tokens[0].replace('(', ' ').trim().intern();
                result[1] = tokens[1].replace(')', ' ').trim().intern();
            }
        }
        return result;
    }

    public boolean containsVertex(String v) {
        if (v != null) {
            return visitedSet.contains(v);
        }
        return false;
    }

    public ArrayList<String> getUnvisitedList() {

        ArrayList<String> list = new ArrayList<String>();
        Iterator<String> it = unvisitedSet.iterator();
        while (it.hasNext()) {
            list.add(it.next());
        }
        return list;
    }

    public void resetUnvisitedSet() {
        this.unvisitedSet = new HashSet<String>();
    }

    public EdgesStruct asEdgesStruct() {
        EdgesStruct edgesStruct = new EdgesStruct();
        edgesStruct.setVertex1(rootVertex);
        edgesStruct.setVertex2("" + getDepth());
        edgesStruct.setElapsedMs(elapsedMs());

        for (int i = 0; i < this.degrees.size(); i++) {
            int level = i + 1;
            JStarDegree jsd = degrees.get(i);
            Iterator<String> vertexIt = jsd.getOutgoingEdgesMap().keySet().iterator();
            while (vertexIt.hasNext()) {
                String v = vertexIt.next();
                Set<DefaultEdge> deSet = jsd.getOutgoingEdgesMap().get(v);
                Iterator<DefaultEdge> edgeIt = deSet.iterator();
                int seq = 0;
                while (edgeIt.hasNext()) {
                    DefaultEdge de = edgeIt.next();
                    String[] vertices = parseDefaultEdge(de);
                    if (vertices.length == 2) {
                        seq++;
                        EdgeStruct es = new EdgeStruct(vertices[0], vertices[1]);
                        es.setSeq(seq);
                        es.setLevel(level);
                        edgesStruct.addEdge(es);
                    }
                }
            }
        }
        return edgesStruct;
    }

    public void display() {
        sysout("JStar Network:");
        sysout("  depth:       " + this.depth);
        sysout("  visitedSize: " + visitedSet.size());
        sysout("  elapsedMs:   " + elapsedMs());
        sysout("  charlotte:   " + visitedSet.contains(PERSON_CHARLOTTE_RAMPLING));
        sysout("  julia:       " + visitedSet.contains(PERSON_JULIA_ROBERTS));
        sysout("  lori:        " + visitedSet.contains(PERSON_LORI_SINGER));
        sysout("  cjoakim:     " + visitedSet.contains("cjoakim"));

        for (int i = 0; i < this.degrees.size(); i++) {
            JStarDegree jsd = degrees.get(i);
            sysout("  degree : " + jsd.getDegree() + " (" + jsd.getOutgoingEdgesMap().size() + ")");
            Iterator<String> vertexIt = jsd.getOutgoingEdgesMap().keySet().iterator();
            while (vertexIt.hasNext()) {
                String v = vertexIt.next();
                sysout("    " + v);
                Set<DefaultEdge> deSet = jsd.getOutgoingEdgesMap().get(v);
                Iterator<DefaultEdge> edgeIt = deSet.iterator();
                int seq = 0;
                while (edgeIt.hasNext()) {
                    seq++;
                    DefaultEdge de = edgeIt.next();
                    //sysout("      " + de);
                }
            }
        }
    }

    public void finish() {
        finishMs = System.currentTimeMillis();
    }

    public long elapsedMs() {
        return finishMs - startMs;
    }

    protected void sysout(String s) {
        System.out.println(s);
    }
}

File: dotnet/src/altgraph_shared_app/Services/Graph/v2/JStarNetwork.cs
using altgraph_shared_app.Services.Graph.v2.Structs;
using QuikGraph;

namespace altgraph_shared_app.Services.Graph.v2
{
  public class JStarNetwork
  {
    public string RootVertex { get; private set; } = string.Empty;
    public int Depth { get; private set; }
    public long StartMs { get; private set; }
    public long FinishMs { get; private set; }
    public List<JStarDegree> Degrees { get; private set; } = new List<JStarDegree>();
    public HashSet<string> VisitedSet { get; private set; } = new HashSet<string>();
    public HashSet<string> UnvisitedSet { get; private set; } = new HashSet<string>();

    public JStarNetwork(string root, int depth)
    {
      this.RootVertex = root;
      this.Depth = depth;
      for (int i = 1; i <= depth; i++)
      {
        Degrees.Add(new JStarDegree(i));
      }
      this.VisitedSet = new HashSet<string>();
      this.UnvisitedSet = new HashSet<string>();
      UnvisitedSet.Add(RootVertex);
    }

    public void AddOutEdgesFor(string vertex, IEnumerable<Edge<string>> outEdges, int degree)
    {
      VisitedSet.Add(vertex);
      Degrees[degree - 1].Add(vertex, outEdges);

      foreach (Edge<string> de in outEdges)
      {
        string[] values = ParseDefaultEdge(de);
        string ev1 = values[0];
        string ev2 = values[1];
        if (!VisitedSet.Contains(ev1))
        {
          UnvisitedSet.Add(ev1);
        }
        if (!VisitedSet.Contains(ev2))
        {
          UnvisitedSet.Add(ev2);
        }
      }
    }

    private string[] ParseDefaultEdge(Edge<string> de)
    {
      string[] result = new string[2];
      result[0] = "?";
      result[1] = "?";

      if (de != null)
      {
        //string[] tokens = de.ToString().Split(":");
        string[] tokens = de.ToString().Split(" -> ");
        if (tokens.Length == 2)
        {
          result[0] = tokens[0].Replace('(', ' ').Trim();
          result[1] = tokens[1].Replace(')', ' ').Trim();
        }
      }
      return result;
    }

    public bool ContainsVertex(string v)
    {
      if (v != null)
      {
        return VisitedSet.Contains(v);
      }
      return false;
    }

    public List<string> GetUnvisitedList()
    {
      List<string> list = new List<string>();
      foreach (string s in UnvisitedSet)
      {
        list.Add(s);
      }
      return list;
    }

    public void ResetUnvisitedSet()
    {
      this.UnvisitedSet = new HashSet<string>();
    }

    public EdgesStruct AsEdgesStruct()
    {
      EdgesStruct edgesStruct = new EdgesStruct();
      edgesStruct.Vertex1 = RootVertex;
      edgesStruct.Vertex2 = "" + Depth;
      edgesStruct.ElapsedMs = ElapsedMs();

      for (int i = 0; i < this.Degrees.Count; i++)
      {
        int level = i + 1;
        JStarDegree jsd = Degrees[i];
        foreach (string v in jsd.OutgoingEdgesMap.Keys)
        {
          IEnumerable<Edge<string>> deSet = jsd.OutgoingEdgesMap[v];
          int seq = 0;
          foreach (Edge<string> de in deSet)
          {
            string[] vertices = ParseDefaultEdge(de);
            if (vertices.Length == 2)
            {
              seq++;
              EdgeStruct es = new EdgeStruct(vertices[0], vertices[1]);
              es.Seq = seq;
              es.Level = level;
              edgesStruct.AddEdge(es);
            }
          }
        }
      }
      return edgesStruct;
    }

    public void Display()
    {
      throw new NotImplementedException();
      // sysout("JStar Network:");
      // sysout("  depth:       " + this.Depth);
      // sysout("  visitedSize: " + VisitedSet.Count);
      // sysout("  elapsedMs:   " + ElapsedMs());
      // sysout("  charlotte:   " + VisitedSet.Contains(ImdbConstants.PERSON_CHARLOTTE_RAMPLING));
      // sysout("  julia:       " + VisitedSet.Contains(ImdbConstants.PERSON_JULIA_ROBERTS));
      // sysout("  lori:        " + VisitedSet.Contains(ImdbConstants.PERSON_LORI_SINGER));
      // sysout("  cjoakim:     " + VisitedSet.Contains("cjoakim"));

      // for (int i = 0; i < this.Degrees.Count; i++)
      // {
      //   JStarDegree jsd = Degrees[i];
      //   sysout("  degree : " + jsd.Degree + " (" + jsd.OutgoingEdgesMap.Count + ")");
      //   Iterator<string> vertexIt = jsd.OutgoingEdgesMap().keySet().iterator();
      //   while (vertexIt.hasNext())
      //   {
      //     string v = vertexIt.next();
      //     sysout("    " + v);
      //     Set<DefaultEdge> deSet = jsd.OutgoingEdgesMap().get(v);
      //     Iterator<DefaultEdge> edgeIt = deSet.iterator();
      //     int seq = 0;
      //     while (edgeIt.hasNext())
      //     {
      //       seq++;
      //       DefaultEdge de = edgeIt.next();
      //       //sysout("      " + de);
      //     }
      //   }
      // }
    }

    public void Finish()
    {
      //FinishMs = System.currentTimeMillis();
    }

    public long ElapsedMs()
    {
      return FinishMs - StartMs;
    }
  }
}